### ISSUE TITLE
Feature request: `scss/function-no-unknown`

### ISSUE THREAD

**Issue by rdebeasi** [stylelint-scss/stylelint-scss#581]
### Problem

Sass doesn't catch typos in function names. The docs [recommend](https://sass-lang.com/documentation/at-rules/function#plain-css-functions) using a linter to catch these issues:

>Because any unknown function will be compiled to CSS, it’s easy to miss when you typo a function name. Consider running a [CSS linter](https://stylelint.io/) on your stylesheet’s output to be notified when this happens!

I'm unable to find any rules in stylelint-scss that catch mistakes like these, however.

### Proposed solution

It would be awesome to have a new rule to catch mistakes like these. It would work like `scss/at-rule-no-unknown`, but for functions. Let's call it `scss/function-no-unknown`. 

### Example

Here's an example of a mistake. I've typed `invart` instead of `invert`:
```scss
@function invert($color, $amount: 100%) {
  $inverse: change-color($color, $hue: hue($color) + 180);
  @return mix($inverse, $color, $amount);
}

.button {
  background-color: invart(#f00)
}
```

Sass assumes that `invart` must be a CSS function and writes it into the compiled CSS:

```css
.button {
  background-color: invart(#f00)
}
```

Our proposed rule would realize that there's no Sass function called `invart` and also no CSS function called `invart`. It would throw an error, alerting us to the fact that we've made a typo.

Thank you so much for your time! :)

**Comment by kristerkari** [comment]
@rdebeasi Thanks for the proposal! I see this as a bit tricky to implement, because `scss/at-rule-no-unknown` is linting for a list of @-rules that are already known.

Functions are something that are created by users, so we don't really know if the function exists in another file, so this rule would really only work if the function is placed in the same file.

**Comment by rdebeasi** [comment]
Totally makes sense.

If there's anything that I can do to help to make this proposal more practical, just let me know. Thank you!

**Comment by gnuletik** [comment]
I'm also looking for a way to avoid these kind of easy mistakes.
I'm currently using the following strategy.

1. Create a second stylelint config file for CSS errors (e.g. `stylelint.dist.config.js`) 
```js
// config to check dist/ folder which is compiled CSS
module.exports = {
  rules: {
    "property-no-unknown": true,
    "media-feature-name-no-unknown": true,
    "at-rule-no-unknown": true,
    "selector-pseudo-class-no-unknown": true,
    "selector-pseudo-element-no-unknown": true,
    "unit-no-unknown": true,

    "function-allowed-list": [
      "scale",
      "scaleY",
      "rgb",
      "rgba",
      "gradient",
      "linear-gradient",
      "color-stop", // inside webkit-linear-gradient
      "calc",
      "translate",
      "translateX",
      "translateY",
      "translateZ",
      "rotate",
      "url",
      "hsla",
      "drop-shadow",
      "rotateX",
      "rotateY",
      "minmax",
      "var",
      "cubic-bezier",
      "saturate",
      "format", // font
      "alpha", // filter
      "counter"
    ]
  }
}
```

2. Edit your build script to check your css after build

```diff
   "scripts": {
-    "build": "vite build",
+    "build": "vite build && npm run check:css",
+    "check:css": "stylelint --config stylelint.dist.config.js 'dist/**/*.css'"
   },
```

However, I did not find a list of known CSS functions so you need to manually add them in your config.

**Comment by ybiquitous** [comment]
Hi there, the Stylelint built-in `function-no-unknown` rule will support the `ignoreFunctions` option with the next version. (stylelint/stylelint#5901)

See also https://github.com/stylelint/stylelint/pull/5865#issuecomment-1034871807

**Comment by kristerkari** [comment]
@ybiquitous looks really nice already. What do you need help with?

**Comment by ybiquitous** [comment]
@kristerkari Thanks. I've left some comments in #591.

**Comment by szaleq** [comment]
> @rdebeasi Thanks for the proposal! I see this as a bit tricky to implement, because `scss/at-rule-no-unknown` is linting for a list of @-rules that are already known.
> 
> Functions are something that are created by users, so we don't really know if the function exists in another file, so this rule would really only work if the function is placed in the same file.

When using Sass `@use` directive, wouldn't it be possible to look for function definitions in the files imported this way?
Example:
```scss
// _functions.scss
@function example() {
    // ..
}

// index.scss
@use "./functions" as fn;

.selector {
    color: fn.example();
}
```
Would it be possible to scan the `_functions.scss` file while linting `index.scss`, list all functions defined in this file and add namespace prefix (`fn` in this example) and then add this list to the list of valid functions for the current file (`index.scss`)?

**Comment by kristerkari** [comment]
> Would it be possible to scan the `_functions.scss` file while linting `index.scss`, list all functions defined in this file and add namespace prefix (`fn` in this example) and then add this list to the list of valid functions for the current file (`index.scss`)?

Doing something like that would most likely make the rule really slow because you would have to be constantly reading the `_functions.scss` file on fly and then parse/lint it. 

Usually going outside of the current file for linting is just a bad idea.
